### PR TITLE
fix(preview): Markdown component layout breaks on copy-paste

### PR DESCRIPTION
**Pull request by Bumbilo**
fixes: https://github.com/openshiftio/openshift.io/issues/3972
- Made the textarea instead of a contenteditable element
screen shot 2019-01-30 at 11 30 19

### CODE REVIEW

File: packages/ngx-widgets/src/app/area-size/area-size.directive.ts [hunk 1]
```diff
@@ -0,0 +1,17 @@
+import { Directive, ElementRef, HostListener, Input, Renderer2 } from '@angular/core';
+
+@Directive({
+  selector: '[textAriaResize]',
```
Comment (christianvogt) [inline packages/ngx-widgets/src/app/area-size/area-size.directive.ts]: Rename to `textAreaResize`

File: packages/ngx-widgets/src/app/area-size/area-size.directive.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+})
+export class AreaSizeDirective {
+  private element: HTMLElement = this.elementRef.nativeElement;
+
+  constructor(private elementRef: ElementRef, private renderer: Renderer2) {}
+
+  @HostListener('input', ['$event'])
+  resizeOnInput() {
+    const input = event.target as HTMLInputElement;
+    this.renderer.setStyle(this.element, 'height', '1px');
```
Comment (christianvogt) [inline packages/ngx-widgets/src/app/area-size/area-size.directive.ts]: Why do you set to `1px` here only to set it to the `scrollHeight` immediately afterwards?

File: packages/ngx-widgets/src/app/markdown/markdown.component.html [hunk 3]
```diff
@@ -17,15 +17,17 @@
     <div class="editor-icon-container">
       <i *ngIf="editAllow" (click)="activeEditor()" class="pficon-edit edit-icon"></i>
     </div>
-    <p
+    <textarea
       almEditable
+      textAriaResize
       #editorInput
+      [style.height.px]="editorInput.scrollHeight"
```
Comment (christianvogt) [inline packages/ngx-widgets/src/app/markdown/markdown.component.html]: What is the purpose of this if you already have the text area resize directive?

File: packages/ngx-widgets/src/app/area-size/area-size.directive.spec.ts [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+  beforeEach(() => {
+    TestBed.configureTestingModule({
+      declarations: [TestComponent, AreaSizeDirective],
+    });
+    fixture = TestBed.createComponent(TestComponent);
+    component = fixture.componentInstance;
+    textAreaElement = fixture.debugElement.query(By.directive(AreaSizeDirective));
+  });
+
+  it('textarea should have equel content height', () => {
```
Comment (christianvogt) [inline packages/ngx-widgets/src/app/area-size/area-size.directive.spec.ts]: `equel` => `equal`

File: packages/ngx-widgets/src/app/area-size/area-size.directive.spec.ts [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    expect(textAreaElement.nativeElement.offsetHeight).toEqual(
+      textAreaElement.nativeElement.scrollHeight,
+    );
+  });
+
+  it('the component should be identified by the directived', () => {
```
Comment (christianvogt) [inline packages/ngx-widgets/src/app/area-size/area-size.directive.spec.ts]: `directived` => `directive`

File: packages/ngx-widgets/src/app/area-size/area-size.directive.spec.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    textAreaElement.triggerEventHandler('input', null);
+    spyOn(textAreaElement.nativeElement, 'focus');
+    expect(textAreaElement).not.toBeNull();
+    fixture.detectChanges();
+    setTimeout(() => {
```
Comment (christianvogt) [inline packages/ngx-widgets/src/app/area-size/area-size.directive.spec.ts]: Is this timeout necessary?
Async tests need to use the `done()` protocol.

File: packages/ngx-widgets/src/app/area-size/area-size.directive.spec.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+  it('textarea should have equel content height', () => {
+    expect(textAreaElement.nativeElement.offsetHeight).toEqual(
+      textAreaElement.nativeElement.scrollHeight,
+    );
+  });
+
+  it('the component should be identified by the directived', () => {
+    textAreaElement.triggerEventHandler('input', null);
+    spyOn(textAreaElement.nativeElement, 'focus');
```
Comment (christianvogt) [inline packages/ngx-widgets/src/app/area-size/area-size.directive.spec.ts]: Why add a focus spy if there's no assertions being made?

File: packages/ngx-widgets/src/app/area-size/area-size.directive.ts (hunk 2 above)
Comment (Bumbilo) [reply packages/ngx-widgets/src/app/area-size/area-size.directive.ts]: When I remove a line in the text area. scrollHeight has the same height and necessary to redefine it.

Comment (christianvogt) [reply packages/ngx-widgets/src/app/area-size/area-size.directive.ts]: Ok I understand why you're doing it now.
I think you need to ensure as well that `overflow-y` is set to `hidden` to account for the scrollbar appearing for OS / browser settings where a scrollbar consumes space. The scrollbar will flash as well as push the text resulting possibly in an incorrect calculation. Test this on several browsers.

I don't know if this is the best solution as it's doing a re-layout of the component on each input change. I can think of other ways to do this but they are certainly more involved with separate height calculations in another element. So if this works and doesn't give any negative experiences to the user (flashing, etc...), i'm ok with it.

**Comment by christianvogt** [comment]
@Bumbilo please test in firefox as I observed editing an existing description loses all new line characters in the text.

**Comment by christianvogt** [comment]
@Bumbilo
Tested in FireFox version:
image

Here's what happens:
markdown_ff

**Comment by Bumbilo** [comment]
@christianvogt  
Fixed
firefox

**Comment by christianvogt** [comment]
Your screen cap shows issues with lists showing up. I tested locally in chrome and FF and experienced the same issue. I cannot create markdown lists anymore.

**Pull request merged**